### PR TITLE
Fix/accessing items view in nested list widget

### LINKED ISSUE #21581: [Bug]: Unable to access third level items (ItemsView)in a List widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

 https://www.loom.com/share/b12807ab1c404c52a9a16c3f4130358d

### Steps To Reproduce

1. DnD a List 
2. Drop another list as child to the first one
3. Drop another Widget inside the nested list

Observe,  From outside the List, you'll observe you cannot access the ItemsView of the second(Nested) List to get the third level widget details

### Public Sample App

_No response_

### Environment

Production

### Issue video log

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

1. Revert the removal of selectedItemView and TriggeredItemView
2. Only remove these properties when child widgets call it.

Fixes #21581

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- Jest
- Cypress

### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 1]
```diff
@@ -179,7 +179,16 @@ enum MODIFICATION_TYPE {
 const ROOT_CONTAINER_PARENT_KEY = "__$ROOT_CONTAINER_PARENT$__";
 const ROOT_ROW_KEY = "__$ROOT_KEY$__";
 const BLACKLISTED_ENTITY_DEFINITION: Record<string, string[] | undefined> = {
-  LIST_WIDGET_V2: ["currentItemsView", "selectedItemView", "triggeredItemView"],
+  LIST_WIDGET_V2: ["currentItemsView"],
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:182]: The issue also exists with the 2nd level of `currentItemsView` when accessed from outside. Can we see if enabling this doesn't cause problems?

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 2]
```diff
@@ -1150,6 +1163,48 @@ class MetaWidgetGenerator {
...
 
+  /**
+   * We can't outrightly remove selectedItemView/triggeredItemView.
+   * List1 --> Level_1
+   *    List2 --> Level_2
+   *       List3 --> Level_3
+   *          Text2
+   *       List4 --> Level_3
+   *          Text3
+   * Assuming we have this setup, Accessing Text2 from Text3 would be
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1174]: Accessing Text2 and Text3 from Level_3 widgets?

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 3]
```diff
@@ -1150,6 +1163,48 @@ class MetaWidgetGenerator {
...
+   */
+
+  private removeBlackListKeysFromEntityDef = (
+    entityDefinition: string,
+    widgetName: string,
+    type: string,
+  ) => {
+    let filteredEntityDef = entityDefinition;
+
+    Object.keys(BLACKLISTED_ENTITY_DEFINITION_IN_LEVEL_DATA).forEach(
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1188]: Any particular reason for loop through the keys rather checking the presence of a value like this?
```
BLACKLISTED_ENTITY_DEFINITION_IN_LEVEL_DATA[type]
```

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 4]
```diff
@@ -1150,6 +1163,48 @@ class MetaWidgetGenerator {
...
+   *       List4 --> Level_3
+   *          Text3
+   * Assuming we have this setup, Accessing Text2 from Text3 would be
+   * {{level_2.currentView.List3.triggeredItemView.Text2.text}}
+   *
+   * So for every widget this.widgetName is the parent List widget name.
+   * And we use this check to selectively blacklist some properties.
+   */
+
+  private removeBlackListKeysFromEntityDef = (
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1181]: Instead of doing this I would prefer to alter the way `getEntityDefinitionsFor` works. 

```
getEntityDefinitionsFor(widgetType: string, blacklistedKeys?: string[])
```

While generating the `currentView` of `levelData` if we find an entry for a particular widget in `BLACKLISTED_ENTITY_DEFINITION_IN_LEVEL_DATA` then we regenerate the entityDefinition for `currentView`.

Doing it like this, may leave opportunities for edge cases.

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1]: Can we add a cypress test for access of `selectedItemView` and `triggeredItemView` of a nested list from page level widgets?

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts (hunk 2 above)
Comment (Tooluloope) [reply app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1174]: Yeah. they are both Level_3 widgets

File: app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts [hunk 5]
```diff
@@ -1150,6 +1162,38 @@ class MetaWidgetGenerator {
...
+   * And we use this check to selectively blacklist some properties.
+   */
+
+  private getFilteredEntityDef = (
+    entityDefinition: string,
+    widgetName: string,
+    type: string,
+  ) => {
+    const filteredEntityDef =
+      this.widgetName === widgetName
```
Comment (ashit-rath) [inline app/client/src/widgets/ListWidgetV2/MetaWidgetGenerator.ts:1186]: We are checking if a child widget has used it's parent list widgets name. But if I have 2 list widgets in level 2, I can now refer one from the other. This will cause cyclic dependency.

Screenshot 2023-03-27 at 11 13 24 AM

I feel the logic here should improve, when using `currentView` through level, all the blacklisted properties are not allowed; atleast for now

**Comment by Aishwarya-U-R** [comment]
> The following are new failures, please fix them before merging the PR:
> 
> 2. cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/ListV2/Listv2_autocomplete_spec.js

@Tooluloope - Since ListV2 is also scripted by you, can we please attempt to fix the flakyness here in this spec in the same PR & then merge it?

**Comment by Aishwarya-U-R** [comment]
Did see one more run recently [here](https://cloud.cypress.io/projects/eyxvp8/runs/31634/test-results) - while browsing for all failing tests in CI runs. I think may be this also comes from this PR runs. Noticing the similarity now.

**Comment by Tooluloope** [comment]
> Did see one more run recently [here](https://cloud.cypress.io/projects/eyxvp8/runs/31634/test-results) - while browsing for all failing tests in CI runs. I think may be this also comes from this PR runs. Noticing the similarity now.

Yes you're correct @Aishwarya-U-R , it's from this same PR.
I'll surely fix them before I merge

**Review by ashit-rath — APPROVED** [review]

**Comment by chandannkumar** [comment]
Tested this PR and LGTM
- Tested if triggeredItemView is accessible
- Tested with all levels of List widget
- Tested on modal
- Tested with older apps
